src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java

```diff
@@ -2695,8 +2695,8 @@ void checkCompatibleSupertypes(DiagnosticPosition pos, Type c) {
         }
         checkCompatibleConcretes(pos, c);
 
-        boolean implementsIdentityObject = types.asSuper(c, syms.identityObjectType.tsym, true) != null;
-        boolean implementsPrimitiveObject = types.asSuper(c, syms.primitiveObjectType.tsym, true) != null;
+        boolean implementsIdentityObject = types.isSubtype(c.referenceProjectionOrSelf(), syms.identityObjectType);
```

Comment:
Not sure you need full blown subtyping here. You are just asking: is this thing derived from "identityObject" - in all these cases asSuper is what you want. Same for AutoCloseable checks. Iterable, ...

Reply:
OK